File: kotlin/src/main/kotlin/io/github/konfigur8/Configuration.kt
package io.github.konfigur8

class Configuration internal constructor(private val settings: Map<Property<*>, String>) : Iterable<Property<*>> {

    /**
     * Retrieve a property value which has been deserialized from it's string form
     * @param prop the property to retrieve the value for
     * @return value of the property
     */
    fun <T> valueOf(prop: Property<T>) = prop.deserialize(
            settings.getOrElse(prop, { throw Misconfiguration("Unknown configuration key '" + prop.name + "'") })
    )

    /**
     * Retrieve a property value which has been deserialized from it's string form
     * @param prop the property to retrieve the value for
     * @return value of the property
     */
    operator fun <T> get(prop: Property<T>) = prop.deserialize(
            settings.getOrElse(prop, { throw Misconfiguration("Unknown configuration key '" + prop.name + "'") })
    )

    /**
     * Public representation of the the settings
     * @return settings described as a map with private values hidden
     */
    fun settings() = settings.map { it.key.name to it.key.exposeMode.display(it.value) }.toMap()

    /**
     * allows you to iterate over property keys
     * @return property key iterator
     */
    override fun iterator(): Iterator<Property<*>> = settings.keys.iterator()

}

File: javascript/src/Configuration.ts
import {Property} from "./Property";
import {UnknownKey} from "./Errors";

export class Configuration {
    private readonly settings: Map<Property<any>, any>;

    constructor(settings: Map<Property<any>, any> = new Map()) {
        this.settings = settings;
    }

    public get<T>(prop: Property<T>) {
        const value = this.settings.get(prop);
        if (!value) {
            throw UnknownKey(prop.name)
        }
        return prop.deserialize(value)
    }
}
